### ISSUE TITLE
UDR hangs if a table presents in both source and target

### ISSUE THREAD

**Issue by crackcraft** [2ndQuadrant/bdr#138]
bdr-plugin/REL0_9_STABLE
version: 0.9.3-2015-10-29-54c8325
postgresql94-server.x86_64       9.4.5-1PGDG.rhel7

db1:

```
  CREATE EXTENSION btree_gist;
  CREATE EXTENSION bdr;
  CREATE TABLE test (id INT, PRIMARY KEY (id));
```

db2:

```
  CREATE EXTENSION btree_gist;
  CREATE EXTENSION bdr;
  CREATE TABLE test (id INT, PRIMARY KEY (id));
  SELECT bdr.bdr_subscribe(local_node_name := 'udr1to2', subscribe_to_dsn := 'host=db1', node_local_dsn := 'host=db2');
  SELECT bdr.bdr_node_join_wait_for_ready();
 -- Hangs here. Replication is not initiated
```

**Comment by ringerc** [comment]
That's expected behaviour at this point in time. If the downstream database is not empty, you need to manually sync the schema and data before subscribing; UDR can only do it for you if the downstream database starts out blank.

You can pass the parameter `synchronize := 'none'` to `bdr.bdr_subscribe` to suppress the initial copy that's done by default.

**Comment by ringerc** [comment]
The reason it appears to hang, rather than give you an error, is that it's necessary to commit the transaction that does the node subscribe before we can know that this will be a problem. You will find more information in the log files on the downstream server if you look at them.

I don't anticipate that this will be an issue that can be solved in the near to medium term. At best we could possibly detect that it'd happen and refuse to subscribe.

The pglogical project that's growing out of UDR has some facilities that I hope will eventually let us solve this in a better way, though right now it behaves much like UDR.

**Comment by crackcraft** [comment]
Thank you a lot. The idea was to use UDR for dumping of roaming databases to a central one.

**Comment by ringerc** [comment]
For that you'll certainly want to suppress the initial dumps then set up the starting state yourself. Keep a close eye out for our pglogical announcement too.

**Comment by akorotkov** [comment]
BTW, what is current status of pglogical project? Is it available somewhere?

**Comment by ringerc** [comment]
The output plugin has been submitted to the commitfest for 9.6:

https://commitfest.postgresql.org/7/418/

The downstream isn't ready to submit yet but is coming along OK. Updates
will be made via the 2ndQuadrant site, pgsql-announce, pgsql-hackers, etc
as appropriate.

I'd love to be able to say more but at the moment there are still some
internal processes outside my control that need to run their course.
​

**Comment by akorotkov** [comment]
OK, thank you for the answer.

**Comment by PJMODOS** [comment]
UDR was superseded by pglogical project so closing this.
